### PR TITLE
Implement recursive copy for gcrane

### PR DESCRIPTION
**Pull request by jonjohnsonjr**
tl;dr `gcrane cp -r gcr.io/foo gcr.io/bar`

This copies images from one repo to another, very quickly.

It's rather hard to test without a fake implementation of GCR, but it works on my machine 😄

### CODE REVIEW

File: pkg/gcrane/copy.go [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+
+	srcAuth authn.Authenticator
+	dstAuth authn.Authenticator
+}
+
+func newCopier(src, dst string) (*copier, error) {
+	srcRepo, err := name.NewRepository(src, name.WeakValidation)
+	if err != nil {
+		if srcRef, refErr := name.ParseReference(src, name.WeakValidation); refErr == nil {
+			srcRepo = srcRef.Context()
```
Comment (mattmoor) [inline pkg/gcrane/copy.go]: Why allow references?

File: pkg/gcrane/copy.go [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+// Workaround for:
+// https://github.com/GoogleCloudPlatform/docker-credential-gcr/issues/54
+type staticAuth struct {
+	auth string
+	err  error
+}
+
+func (sa staticAuth) Authorization() (string, error) {
+	return sa.auth, sa.err
+}
```
Comment (mattmoor) [inline pkg/gcrane/copy.go]: Yuck

File: pkg/gcrane/copy.go [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+	dstRepo, err := name.NewRepository(dst, name.WeakValidation)
+	if err != nil {
+		if dstRef, refErr := name.ParseReference(dst, name.WeakValidation); refErr == nil {
+			dstRepo = dstRef.Context()
+		} else {
+			return nil, fmt.Errorf("parsing repo %q: %v", dst, err)
+		}
+	}
+
+	srcHelper, err := authn.DefaultKeychain.Resolve(srcRepo.Registry)
```
Comment (mattmoor) [inline pkg/gcrane/copy.go]: Why does gcrane use the default keychain?  Surely we can do better.

File: pkg/gcrane/copy.go (hunk 2 above)
Comment (jonjohnsonjr) [reply pkg/gcrane/copy.go]: Yeah :/

/cc @dekkagaijin want to make me a caching cred helper and/or `authn.Authenticator`? 😃

Comment (dekkagaijin) [reply pkg/gcrane/copy.go]: Lets talk after kubecon? :sob:

File: pkg/gcrane/copy.go (hunk 1 above)
Comment (jonjohnsonjr) [reply pkg/gcrane/copy.go]: To share some of the parsing/auth code. Moved some stuff around to make it a little better.

File: pkg/gcrane/copy.go (hunk 3 above)
Comment (jonjohnsonjr) [reply pkg/gcrane/copy.go]: Okay we actually can do better, this makes things a lot faster for listing without having to do the stupid static auth thing.

https://github.com/google/go-containerregistry/pull/315

**Comment by jonjohnsonjr** [comment]
@mattmoor I rebased this on #315, should be RFAL.

File: pkg/gcrane/copy.go [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+
+	srcAuth, err := authn.DefaultKeychain.Resolve(srcRef.Context().Registry)
+	if err != nil {
+		return nil, nil, fmt.Errorf("getting auth for %q: %v", src, err)
+	}
+
+	dstAuth, err := authn.DefaultKeychain.Resolve(dstRef.Context().Registry)
+	if err != nil {
+		return nil, nil, fmt.Errorf("getting auth for %q: %v", dst, err)
+	}
```
[the remark this answers is not in the document]
Comment (jonjohnsonjr) [reply pkg/gcrane/copy.go]: Don't tempt me!

**Pull request merged**